### PR TITLE
Optimize Pipe.DefaultPipeWriter GetSpan and Advance

### PR DESCRIPTION
**Pull request by ahsonkhan**
- Memory.Slice.Span is slightly slower than Memory.Span.Slice (GetSpan helper gets the span first before slicing).
- Removing the second (length) parameter from Slice since it isn't necessary.
- I am assuming that Advance(bytesWritten = 0) is not a common scenario, so we shrink the method size by not creating a separate code path for it and save a branch.

cc @pakrym, @davidfowl, @benaadams

### CODE REVIEW

File: src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs [hunk 1]
```diff
@@ -135,76 +135,112 @@ internal Memory<byte> GetMemory(int sizeHint)
...
+            lock (_sync)
+            {
+                AllocateWriteHeadUnsynchronized(sizeHint);
             }
 
             // Slice the AvailableMemory to the WritableBytes size
             int end = _writingHead.End;
-            Memory<byte> availableMemory = _writingHead.AvailableMemory;
-            availableMemory = availableMemory.Slice(end, availableMemory.Length - end);
+            Span<byte> availableMemory = _writingHead.AvailableMemory.Span;
```
Comment (benaadams) [inline src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs]: nit: `availableMemory` => `availableSpan`

**Comment by pakrym** [comment]
Any perf numbers?

File: src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs (hunk 1 above)
Comment (pakrym) [reply src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs]: Don't use `_writingHead` outside the lock

Comment (benaadams) [reply src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs]: Was used previously outside lock?

Comment (pakrym) [reply src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs]: Good time to stop :)

Comment (pakrym) [reply src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs]: Or wait, it's not supposed to be syncronised.

**Comment by ahsonkhan** [comment]
> Any perf numbers?

I don't know of any relevant benchmarks that exist today that will help with this (at this scale). Any ideas which ones to run and compare? Nano benchmarks, measuring each of the methods in isolation show small improvements (5-10%).

``` ini

BenchmarkDotNet=v0.10.14, OS=Windows 10.0.16299.431 (1709/FallCreatorsUpdate/Redstone3)
Intel Xeon CPU E5-1620 v2 3.70GHz, 1 CPU, 8 logical and 4 physical cores
Frequency=3604598 Hz, Resolution=277.4234 ns, Timer=TSC
.NET Core SDK=2.1.300-rtm-008866
  [Host]     : .NET Core 2.1.0-rtm-26516-03 (CoreCLR 4.6.26516.03, CoreFX 4.6.26516.03), 64bit RyuJIT
  DefaultJob : .NET Core 2.1.0-rtm-26516-03 (CoreCLR 4.6.26516.03, CoreFX 4.6.26516.03), 64bit RyuJIT
```

**If marked with AggressiveInlining attribute:**

|          Method |     Mean |     Error |    StdDev | Scaled | ScaledSD |
|---------------- |---------:|----------:|----------:|-------:|---------:|
|   GetMemorySpan | 3.728 ns | 0.0300 ns | 0.0234 ns |   1.00 |     0.00 |
| GetSpanDirectly | 3.254 ns | 0.0973 ns | 0.1121 ns |   0.87 |     0.03 |

**If not inlined:**

|          Method |      Mean |     Error |    StdDev | Scaled |
|---------------- |----------:|----------:|----------:|-------:|
|   GetMemorySpan | 13.275 ns | 0.0895 ns | 0.0699 ns |   1.00 |
| GetSpanDirectly |  4.238 ns | 0.1177 ns | 0.1488 ns |   0.32 |

|                 Method |     Mean |     Error |    StdDev | Scaled | ScaledSD |
|----------------------- |---------:|----------:|----------:|-------:|---------:|
| GetSegmentSizeOriginal | 2.926 ns | 0.0296 ns | 0.0262 ns |   1.00 |     0.00 |
|      GetSegmentSizeNew | 2.865 ns | 0.0571 ns | 0.0506 ns |   0.98 |     0.02 |

**Pull request merged**